rust/automerge-c/src/result.rs

```diff
@@ -184,13 +184,20 @@ impl From<(&am::Value<'_>, &RefCell<Option<CString>>)> for AMvalue<'_> {
                 am::ScalarValue::Null => AMvalue::Null,
                 am::ScalarValue::Str(smol_str) => {
                     let mut c_str = c_str.borrow_mut();
-                    AMvalue::Str(match c_str.as_mut() {
+                    match c_str.as_mut() {
                         None => {
-                            let value_str = CString::new(smol_str.to_string()).unwrap();
-                            c_str.insert(value_str).as_ptr()
+                            let value_str = smol_str.as_str();
+                            if value_str.find('\0').is_some() {
+                                // A C string can't contain null characters
+                                // so we have to treat the string as bytes.
+                                AMvalue::Bytes(value_str.as_bytes().into())
```

Comment:
What happens in the following scenario:

- Bob creates an automerge document in JavaScript with a string property (let's call it `"badname"`) which contains a null byte
- Bob saves the document and sends it to Alice
- Alice loads the document using the C library, loads the `"badname"` property and adds a byte to the end
- Alice saves the document and sends it back to Bob
- Bob loads the document in Java-script again

From Bob's perspective the `"badname"` property has now changed from `String` to `Bytes` right?

Reply:
@alexjg, from Alice's perspective, the "badname" value has effectively become read-only in the same sense that a value tagged as being `AM_VALUE_UNKNOWN` is read-only because it cannot be translated into an analogous C type.

As you rightly pointed out, the problem with this approach is that it doesn't communicate the value's translation into an alternate type and, even worse, Alice isn't prevented from editing it either.

It appears that a better approach would be to translate the value into an `AM_VALUE_UNKNOWN` instead because its bytes can still be read through the `bytes` member, its true data type can be communicated through the `type_code` member and the recipient will be unable to edit it.

@alexjg, what value should the `type_code` member be in this situation?

Reply:
Well, the type code would be `6` as specified [here](https://alexjg.github.io/automerge-storage-docs/#raw-value-columns).

However, it feels like the real problem is that we are using `CString` for the type of `AMresult::String` when in fact strings in automerge are UTF8 (which can contain null bytes). Should we consider using just a `Vec<u8>` there instead?

Reply:
I like the idea of representing an AMresult::String using the same structure as an AMbyteSlice (and I share @alexjg's concern with having the type of the value change depending on its content).

If we want to support nulls in map keys too then that will require a major (if mechanical) change to the API to pass a pointer + length instead of a Cstring for `AMmapPut*` too.

The other APIs that ned fixing are `AMmapItemsValue`, `AMlistItemsValue`, `AMobjItemValue` (though I am not currently using these for automerge-go due to the friction around reading the AMvalue union using cgo).

Reply:
> Well, the type code would be 6 as specified [here](https://alexjg.github.io/automerge-storage-docs/#raw-value-columns).

@alexjg, thank you!

> However, it feels like the real problem is that we are using CString for the type of AMresult::String when in fact strings in automerge are UTF8 (which can contain null bytes).

Reading that I'm reminded of Frank Zappa's quote about curing dandruff with decapitation.

I think it's fair to say that %99+ of strings in human-generated content aren't going to contain null bytes because the vast majority of physical and virtual keyboards do not furnish a NUL key; more than one step is required to insert a NUL character into a text document.

Isn't its continued use as a string-terminating sentinel in what would otherwise be just an array of bytes the chief reason for the inclusion of NUL (U+0000) in the Unicode character set? It is, after all, categorized as a control character.

It seems to me that if one wishes to store an array of bytes in an Automerge document then one should do so explicitly as a bytes value and not as a string. Translating an array of bytes that was conveyed as a string into an immutable array of bytes that's identified as a string provides a group of collaborators with an opportunity to negotiate how the value should be stored in future revisions of the document.

Reply:
> The other functions that need fixing are `AMmapItemValue()`, `AMlistItemValue()`, `AMobjItemValue()`...

@ConradIrwin, I believe those functions are covered by the changes I've made already but I should add unit tests to confirm.

Reply:
I agree that putting a null byte in the middle of a UTF-8 string is a strange thing to do. However, we have specified that string values in automerge are UTF-8, so I think we should gracefully handle users doing such unnatural things. Using `Unknown` might work - we would need to add a lot of documentation to the effect that if you are using the library via FFI from an environment which does allow `U+0000` then you must handle both `Amresult::String` and `Amresult::Unknown`. This seems like it might be quite confusing.

It might well be better to just return an `Amresult::Error` here as @ConradIrwin suggests.

Reply:
> It might well be better to just return an Amresult::Error here as @ConradIrwin suggests.

@alexjg, I suspect that in most cases receiving an `AM_VALUE_UNKNOWN` will manifest as an error anyway because the application's assertion that the `AMvalue.tag` member's value is `AM_VALUE_STR` will have been violated.

Returning a (usually) unexpected value instead of an error gives an organization the option of migrating legacy documents from a schema that forces a bytes field to be stored as a string to a schema where they're explicitly stored as bytes or to remove/substitute NUL characters in string values that were stored prior to the institution of a policy that forbids NUL characters within string values.

Reply:
On the flip side errors have descriptive error messages... it would be quite confusing to get an `AMunknown` and not know why.

Reply:
I still think we should either handle UTF-8 strings throughout, or throw an error at this point. To me having a field which changes type depending on it's contents but which appears the same to the user seems quite confusing.

Note that I don't think we should assume that this is a case of someone attempting to store bytes in a string - it's equally likely that it happens as a bug in some text editor somewhere or as a weird escape sequence in source code or any number of other situations which produce some cursed utf8. In any situation where the user creating the document is working in a utf8 environment (e.g. source code of a program creating automerge documents, or a text editor, or a JavaScript console etc. etc.) the user would have no reason to believe there was anything wrong until the document causes someone loading it with the C library to get an `Unknown` value, this seems like it would be very hard to debug. On the flip side an error saying "this library doesn't handle utf8 with null bytes" is quite a lot clearer.

Imagine in a years time someone comes to us with a bug report from an FFI wrapper of `automerge-c` in Ruby saying "this document keeps telling me there's an unexpected Unknown type" vs "when I load this document it tells me it can't handle null byte utf8", which would be easier to help them debug? :slightly_smiling_face:

Reply:
> On the flip side errors have descriptive error messages... it would be quite confusing to get an AMunknown and not know why.

@ConradIrwin, I think there's only one reason why an `AM_VALUE_UNKNOWN` was returned instead of the expected `AM_VALUE_STR`: there's more than one legitimate way to convert—or decline to convert—the underlying UTF-8 string into a `char const*` and the automerge-c library is refusing the temptation to guess.

As you've pointed out, some higher-level languages may not be able to access the members of the `AMvalue` union directly so automerge-c should provide optional functions for accessing those members as if `AMvalue` were an opaque type.

The `AMvalueStr()` function should take a callback function argument so that an application can specify on a case-by-case basis how a returned `AM_VALUE_UNKNOWN` should be handled either by conversion or error reporting. The function should also take a `void*` in/out argument so that state can be passed through to the callback function as well e.g. an error-reporting callback function can update an application-level error variable for languages that don't support exceptions.

For an organization that forbids NUL characters from being inserted into a string value, `AMvalue.str` can be accessed directly but for an organization that can't be as certain then they should call `AMvalueStr()` with an appropriate callback function pointer and state variable pointer instead.

Reply:
The reason for the `unknown` type code is to allow for us to add types in future and still allow older implementations to work with newer files containing the new types. This means that there is another reason that someone might get `AM_VALUE_UNKNOWN` instead of `AM_VALUE_STR`: if someone with a new version of automerge set the given property to a new type.  I suppose that could still be handled with the approach you're suggesting by switching on the `type_code`?

Reply:
> This means that there is another reason that someone might get AM_VALUE_UNKNOWN instead of AM_VALUE_STR: if someone with a new version of automerge set the given property to a new type. I suppose that could still be handled with the approach you're suggesting by switching on the type_code?

@alexjg, yes, the callback function can use the `AMunknownValue.type_code` member to disambiguate in that circumstance.

All of the remaining `AMvalue*()` access functions should resemble the `AMvalueStr()` function that I've proposed so that they, too, can be flexibly configured to coerce unknown types into known ones as appropriate.

Reply:
Back to where we started, should AMvalueStr() just always return an AMbyteSpan?

Almost all languages internally represent a string as a pointer + a length to avoid the obvious problems with the C representation (and it would be no harder for automerge-go to use this representation to convert to a Go string).

That way anyone who is using automerge-c directly will have to deal with it; but anyone who's just using it as a common translation layer will not have to worry. I suspect (though I have an obvious bias :D) there will be more users using automerge-c as FFI not directly.

Reply:
> Back to where we started, should AMvalueStr() just always return an AMbyteSpan?

@ConradIrwin, after more consideration this is the approach towards which I'm leaning due to its simplicity.

For me, anything is preferable to the error approach because it would mean that applications using automerge-c wouldn't be able to process every Automerge document.

> I suspect (though I have an obvious bias :D) there will be more users using automerge-c as FFI not directly.

You're probably right about that.

Reply:
I like the string as `AMbyteSpan` approach. There is still one consideration which is that if you put non utf8 bytes in a string field then  it will not be loadable as the decoding code checks that the bytes are valid utf8. This won't result in unloadable documents because `automerge_c::doc::utils::to_str` would null out the string before putting it in the document, but it would be a little surprising. We would probably want to modify `AMmapPutStr` to return an error for non utf8 bytes in a string.

Reply:
Sorry that's slightly incorrect, `to_str` would currently call `Cstring::to_string_lossy`. The equivalent for rust strings would be `String::from_utf8_lossy`.

Reply:
makes sense!

I think this is a great solution for values, and I like the utf8 check (similar to rust go's strings are utf-8 "by convention"). 

What do we want to do about map keys?
* Would it be reasonable to check for and disallow `U+0000` in map keys? If so, that seems the easiest for everyone.
* If not... should we make the same change so that map keys are represented as `AMbyteSpan` too; or just error?

Reply:
> If not... should we make the same change so that map keys are represented as `AMbyteSpan` too; or just error?

@ConradIrwin, I think that string values should be handled consistently and that an application based on automerge-c should be able to process  an Automerge document with NUL characters in its map keys, too.